File: backend/src/main/java/com/meditru/controller/GeminiController.java
package com.meditru.controller;

import com.meditru.dto.*;
import com.meditru.service.GeminiService;
import com.meditru.util.Validators;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;

@RestController
@RequestMapping("/api/gemini")
public class GeminiController {

    private final GeminiService geminiService;

    public GeminiController(GeminiService geminiService) {
        this.geminiService = geminiService;
    }

    @PostMapping("/health-assistant")
    public ResponseEntity<?> healthAssistant(@RequestBody HealthAssistantRequest request) {
        // Validate message
        if (!Validators.isNonEmptyString(request.message(), 5000)) {
            return ResponseEntity.badRequest().body(
                ApiErrorResponse.of("Invalid input", "'message' must be a non-empty string (max 5000 chars).")
            );
        }

        String cleanMessage = Validators.sanitize(request.message(), 5000);
        String cleanReportContext = Validators.isNonEmptyString(request.reportContext(), 10000)
            ? Validators.sanitize(request.reportContext(), 10000)
            : null;

        // Sanitize history
        List<ChatHistoryEntry> cleanHistory = new ArrayList<>();
        if (request.history() != null) {
            List<ChatHistoryEntry> recent = request.history().size() > 10
                ? request.history().subList(request.history().size() - 10, request.history().size())
                : request.history();
            for (ChatHistoryEntry entry : recent) {
                cleanHistory.add(new ChatHistoryEntry(
                    Validators.sanitize(entry.role(), 50),
                    Validators.sanitize(entry.text(), 2000)
                ));
            }
        }

        HealthAssistantResponse response = geminiService.generateHealthResponse(
            cleanMessage, cleanReportContext, cleanHistory
        );
        return ResponseEntity.ok(response);
    }

    @PostMapping("/clinical-notes")
    public ResponseEntity<?> clinicalNotes(@RequestBody ClinicalNotesRequest request) {
        // Validate patientName
        if (!Validators.isNonEmptyString(request.patientName(), 200)) {
            return ResponseEntity.badRequest().body(
                ApiErrorResponse.of("Invalid input", "'patientName' must be a non-empty string (max 200 chars).")
            );
        }

        ClinicalNotesRequest cleanRequest = new ClinicalNotesRequest(
            Validators.sanitize(request.patientName(), 200),
            (request.age() != null && request.age() > 0 && request.age() < 150) ? request.age() : null,
            Validators.isNonEmptyString(request.symptoms(), 5000) ? Validators.sanitize(request.symptoms(), 5000) : null,
            Validators.isNonEmptyString(request.vitals(), 1000) ? Validators.sanitize(request.vitals(), 1000) : null,
            Validators.isNonEmptyString(request.consultationTranscript(), 10000)
                ? Validators.sanitize(request.consultationTranscript(), 10000) : null
        );

        ClinicalNotesResponse response = geminiService.generateClinicalNotes(cleanRequest);
        return ResponseEntity.ok(response);
    }
}

File: backend/src/controllers/gemini.controller.ts
import { Request, Response } from "express";
import { sanitize, isNonEmptyString } from "../utils/validators";
import { generateHealthResponse, generateClinicalNotes } from "../services/gemini.service";

export async function healthAssistant(req: Request, res: Response): Promise<void> {
  try {
    const { message, history = [], reportContext } = req.body;

    if (!isNonEmptyString(message, 5000)) {
      res.status(400).json({
        error: "Invalid input",
        message: "'message' must be a non-empty string (max 5000 chars).",
      });
      return;
    }

    const cleanMessage = sanitize(message, 5000);
    const cleanReportContext =
      reportContext && typeof reportContext === "string"
        ? sanitize(reportContext, 10000)
        : undefined;

    const cleanHistory = Array.isArray(history)
      ? history
          .slice(-10)
          .map((h: { role?: string; text?: string }) => ({
            role: typeof h.role === "string" ? sanitize(h.role, 50) : "unknown",
            text: typeof h.text === "string" ? sanitize(h.text, 2000) : "",
          }))
      : [];

    const result = await generateHealthResponse(cleanMessage, cleanReportContext, cleanHistory);
    res.json(result);
  } catch (error: unknown) {
    const errMsg = error instanceof Error ? error.message : "Unknown error";
    console.error(`[ERROR] /api/gemini/health-assistant: ${errMsg}`);
    res.status(500).json({
      error: "Failed to generate health advice",
      message: "Internal AI server error. Please try again later.",
    });
  }
}

export async function clinicalNotes(req: Request, res: Response): Promise<void> {
  try {
    const { patientName, age, symptoms, vitals, consultationTranscript } = req.body;

    if (!isNonEmptyString(patientName, 200)) {
      res.status(400).json({
        error: "Invalid input",
        message: "'patientName' must be a non-empty string (max 200 chars).",
      });
      return;
    }

    const result = await generateClinicalNotes({
      patientName: sanitize(patientName, 200),
      age: typeof age === "number" && age > 0 && age < 150 ? age : undefined,
      symptoms: isNonEmptyString(symptoms, 5000) ? sanitize(symptoms, 5000) : undefined,
      vitals: isNonEmptyString(vitals, 1000) ? sanitize(vitals, 1000) : undefined,
      consultationTranscript: isNonEmptyString(consultationTranscript, 10000)
        ? sanitize(consultationTranscript, 10000)
        : undefined,
    });

    res.json(result);
  } catch (error: unknown) {
    const errMsg = error instanceof Error ? error.message : "Unknown error";
    console.error(`[ERROR] /api/gemini/clinical-notes: ${errMsg}`);
    res.status(500).json({
      error: "Failed to generate clinical notes",
      message: "Internal AI server error. Please try again later.",
    });
  }
}
